bot/wikidata/federation_tester.py

```diff
@@ -46,49 +51,52 @@ def run(self):
-        newtext += u'{| class="wikitable sortable"\n! Link !! Git !! Wiki !! Query works?\n'
+        reportpage = pywikibot.Page(self.repo, title='Wikidata:SPARQL query service/Federation report')
+        newtext = 'Testing the different SPARQL endpoints. Taking the links from '
+        newtext += '[https://gerrit.wikimedia.org/r/plugins/gitiles/operations/puppet/+/refs/heads/production/modules/query_service/templates/allowlist.txt.epp git] and '
+        newtext += '[[:mw:Wikidata Query Service/User Manual/SPARQL Federation endpoints|the manual]] and doing a '
+        newtext += '[https://query.wikidata.org/#SELECT%20%3Fa%20%3Fb%20%3Fc%20WHERE%20%7B%20%0A%20%20SERVICE%20%3Chttp%3A%2F%2Fdata.bibliotheken.nl%2Fsparql%3E%20%7B%0A%20%20%20%20SELECT%20%3Fa%20%3Fb%20%3Fc%20WHERE%20%7B%20%3Fa%20%3Fb%20%3Fc%20%7D%20LIMIT%201%0A%20%20%7D%0A%7D%20LIMIT%201 very simple federated query] ([https://github.com/multichill/toollabs/blob/master/bot/wikidata/federation_tester.py bot source code]).\n\n'
+        newtext += '{| class="wikitable sortable"\n! Link !! Git !! Wiki !! Query works?\n'
         alllist = sorted(set(self.gitlist) | set(self.wikilist))
         for link in alllist:
-            newtext += u'|-\n'
-            newtext += u'| %s\n' % (link,)
+            newtext += '|-\n'
+            newtext += '| %s\n' % (link,)
             if link in self.gitlist:
-                newtext += u'| OK\n' # FIXME: Add pretty icon
+                newtext += '| OK\n' # FIXME: Add pretty icon
             else:
-                newtext += u'| Missing\n'
+                newtext += '| Missing\n'
             if link in self.wikilist:
-                newtext += u'| OK\n' # FIXME: Add pretty icon
+                newtext += '| OK\n' # FIXME: Add pretty icon
             else:
-                newtext += u'| Missing\n'
+                newtext += '| Missing\n'
             queryworks = self.testSparql(link)
             if queryworks:
-                newtext += u'| OK\n' # FIXME: Add pretty icon
+                newtext += '| OK\n' # FIXME: Add pretty icon
             else:
-                newtext += u'| Query failed\n'
-        newtext += u'|}\n[[Category:Wikidata:SPARQL query service]]'
-        print newtext
-        reportpage.put(newtext, summary=u'Updating federation report')
+                newtext += '| Query failed\n'
+        newtext += '|}\n[[Category:Wikidata:SPARQL query service]]'
+        print(newtext)
+        reportpage.put(newtext, summary='Updating federation report')
 
     def testSparql(self, link):
-        query = u"""SELECT ?a ?b ?c WHERE { 
+        query = """SELECT ?a ?b ?c WHERE { 
   SERVICE <%s> {
     SELECT ?a ?b ?c WHERE { ?a ?b ?c } LIMIT 1
   }
 } LIMIT 1""" % (link,)
+        headers = pywikibot.data.sparql.DEFAULT_HEADERS | {
```

Comment:
Pywikibot is already setting a (much better) user-agent, see https://www.mediawiki.org/wiki/Manual:Pywikibot/User-agent . If that's not the case for the SPARQL endpoint, that would be a bug

Reply:
I only looked at `sparql.py` and didn’t see a User-Agent there – looks like it’s buried in `http.py`. I’ll remove it again.

Reply:
(Though I’m not convinced by “much better”, as the user agent will presumably include pywikibot information but be lacking the specific `federation_tester.py` component. But as it doesn’t look like the sparql module lets us control the user agent very much, it’s probably still better to leave it at the default.)

Reply:
"federation_tester" will be the script name included in the user-agent